### PR TITLE
Allow compositional fields to be grouped as vectors in visualization

### PR DESCRIPTION
**Pull request by jperryhouts**
Alternative to #541

### CODE REVIEW

**Comment by bangerth** [comment]
I think I'd rather have the list of compositional fields listed in the input file, rather than using magic names. What do you think?

**Comment by jperryhouts** [comment]
Fixed. This plugin takes a list of compositional field names in the format you suggested (vec1x, vec1y; vec2x, vec2y) and a list of names for each output field. The only thing that would be nice to add is the ability to avoid redundantly outputting those fields as scalars, but that's not such a big deal to leave as is.

File: source/postprocess/visualization/compositional_vector.cc [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+  namespace Postprocess
+  {
+    namespace VisualizationPostprocessors
+    {
+      ASPECT_REGISTER_VISUALIZATION_POSTPROCESSOR(CompositionalVector,
+                                                  "compositional vector",
+                                                  "A visualization output object that outputs vectors whose "
+                                                  "components are derived from compositional fields. Input "
+                                                  "parameters for this postprocessor are defined in section "
+                                                  "Postprocess.Visualization.Compositional fields as vectors")
```
Comment (tjhei) [inline source/postprocess/visualization/compositional_vector.cc]: we typically write paths of subsections with a forward slash (so "Postprocess/Visualization/Compositional fields").

**Comment by tjhei** [comment]
Is it useful to have the option to pick non-consecutive compositional names as belonging to a single vector? If not, it would be simpler to ask the user to just pick the first index. Thoughts?

Looks good otherwise. /run-tests

**Comment by jperryhouts** [comment]
Good idea. I added the option to only request one compositional field in a set, in which case it is interpreted as the first in a consecutive list. The more general functionality is still available.

**Comment by bangerth** [comment]
Nice. I'm not sure what to do about the testcase given that there is no p4est version we can currently use.  @tamiko -- is there a way to condition the test on having at least p4est 1.2 (whenever that comes out)?

**Comment by bangerth** [comment]
Sorry, the previous comment was meant for another PR.

File: source/postprocess/visualization/compositional_vector.cc [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+      template <int dim>
+      std::vector<std::string>
+      CompositionalVector<dim>::
+      get_names () const
+      {
+        std::vector<std::string> solution_names;
+        for (unsigned int i=0; i<vector_names.size(); ++i)
+          for (unsigned int j=0; j<dim; ++j)
+            solution_names.push_back(vector_names[i]);
```
Comment (bangerth) [inline source/postprocess/visualization/compositional_vector.cc]: You're creating a vector of strings by pushing back the elements of a vector of strings. You could just return the original vector of strings here :-)

Comment (jperryhouts) [inline source/postprocess/visualization/compositional_vector.cc]: Nope. That skips the `j` loop. Each string from vector_names needs to be repeated `dim` times in the final vector. Maybe there's a more clever c++ way to do that, though?

Comment (bangerth) [inline source/postprocess/visualization/compositional_vector.cc]: Ah, oh, indeed. Thanks for pointing it out.

File: source/postprocess/visualization/compositional_vector.cc [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+      template <int dim>
+      std::vector<DataComponentInterpretation::DataComponentInterpretation>
+      CompositionalVector<dim>::
+      get_data_component_interpretation () const
+      {
+        std::vector<DataComponentInterpretation::DataComponentInterpretation> interpretation;
+        for (unsigned int i=0; i<vector_names.size()*dim; ++i)
+          interpretation.push_back (DataComponentInterpretation::component_is_part_of_vector);
+
+        return interpretation;
```
Comment (bangerth) [inline source/postprocess/visualization/compositional_vector.cc]: You can write this shorter as
```
  return std::vector<DataComponentInterpretation::DataComponentInterpretation>                 
              (DataComponentInterpretation::component_is_part_of_vector,
               vector_names.size()*dim);
```

File: source/postprocess/visualization/compositional_vector.cc [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+        for (unsigned int q=0; q<n_quadrature_points; ++q)
+          {
+            std::vector<double> composition(this->n_compositional_fields());
+
+            for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+              composition[c] = uh[q][this->introspection().component_indices.compositional_fields[c]];
+
+            for (unsigned int i=0; i<vector_names.size(); ++i)
+              for (unsigned int j=0; j<dim; ++j)
+                computed_quantities[q][i*dim+j] = composition[sets[i][j]];
```
Comment (bangerth) [inline source/postprocess/visualization/compositional_vector.cc]: This seems inefficient. You first copy *all* components from `uh[q]` into the `composition` array, and then you subset this field into `computed_quantities`. I think you could do this in one step without the copy operation.

If you think it's clearer the way you have it right now, please pull the declaration of the `composition` array out of the loop.

File: source/postprocess/visualization/compositional_vector.cc [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+            prm.enter_subsection("Compositional fields as vectors");
+            {
+              prm.declare_entry("Names of vectors", "",
+                                Patterns::List (Patterns::Anything()),
+                                "Names of vectors as they will appear in the output.");
+
+              prm.declare_entry("Names of fields", "",
+                                Patterns::Anything (),
+                                "A list of sets of compositional fields which should be output "
+                                "as vectors. Sets are separated from eachother by semicolons "
```
Comment (bangerth) [inline source/postprocess/visualization/compositional_vector.cc]: eachother -> each other

File: source/postprocess/visualization/compositional_vector.cc [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+          prm.leave_subsection();
+        }
+        prm.leave_subsection();
+      }
+
+
+      template <int dim>
+      void
+      CompositionalVector<dim>::parse_parameters (ParameterHandler &prm)
+      {
```
Comment (bangerth) [inline source/postprocess/visualization/compositional_vector.cc]: nice work in this function doing complete error checking!

**Comment by bangerth** [comment]
Nice work. Just a couple of nitpicks. Let me know when you've addressed them and we can merge.

**Comment by jperryhouts** [comment]
OK. I think it's good now if the tester is okay with it.

**Comment by bangerth** [comment]
Oh, I merged, and then promptly realized that the patch does not have an addition to the changed.h file. Can you write a pull request that just announces the new postprocessor?

**Comment by jperryhouts** [comment]
Oh, yep. I'm in the field right now but I'll do it as soon as I'm back (on Tuesday evening).

**Pull request merged**